recursion/core/src/poseidon2/external.rs

```diff
@@ -50,56 +71,90 @@ impl<F: PrimeField32> MachineAir<F> for Poseidon2Chip {
     }
 
     #[instrument(name = "generate poseidon2 trace", level = "debug", skip_all)]
+    #[allow(clippy::needless_range_loop)]
     fn generate_trace(
         &self,
-        _: &ExecutionRecord<F>,
-        _: &mut ExecutionRecord<F>,
+        input: &ExecutionRecord<F>,
+        _output: &mut ExecutionRecord<F>,
     ) -> RowMajorMatrix<F> {
-        let mut input = [F::one(); WIDTH];
-        let rows = (0..128)
-            .map(|i| {
+        let mut rows = Vec::new();
+        // let mut record = ExecutionRecord::default();
+
+        for i in 0..input.poseidon2_events.len() {
+            let event = input.poseidon2_events[i].clone();
+
+            // init state for the first round.
+            // this will also be used to store the pre-state for the next round.
+            let mut pre_state = event.initial_state;
+
+            let memory = event.initial_state_records.clone();
+
+            for i in 0..NUM_TOTAL_ROUNDS {
                 let mut row = [F::zero(); NUM_POSEIDON2_COLS];
                 let cols: &mut Poseidon2Cols<F> = row.as_mut_slice().borrow_mut();
-                cols.input = input;
 
-                let r = i % 31;
-                let rounds_f = 8;
-                let rounds_p = 22;
-                let rounds = rounds_f + rounds_p;
-                let rounds_f_beginning = rounds_f / 2;
-                let p_end = rounds_f_beginning + rounds_p;
+                // increment the clock.
```

Comment:
please keep commenting style consistent with the rest of the repo.

Full sentences, proper capitalization, periods